src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,142 @@
+            self.log.warn("Failed to fetch MDS service details for {}"
+                           .format(fs_name))
+            return None
+
+        svc = svclist[0]
+        self.log.info("newsvc:{}".format(svc))
+        self.log.info("newsvc.spec:{}".format(svc.spec))
+        self.log.info("new count:{}".format(abscount))
+
+        hosts = []
+        for h in svc.spec.placement.hosts:
+            hosts.append(h.hostname)
+        ps = PlacementSpec(hosts=hosts, count=abscount)
+        newsvc = ServiceSpec(service_type=svc.spec.service_type,
+                             service_id=svc.spec.service_id,
+                             placement=ps)
+        return newsvc
+
+    def spawn_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count+1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def kill_mds(self, fs_name, count):
+        newsvc = self.update_daemon_count(fs_name, count-1)
+        if newsvc is None:
+            return
+
+        completion = self.apply_mds(newsvc)
+        self._orchestrator_wait([completion])
+        orchestrator.raise_if_exception(completion)
+
+    def get_required_standby_count(self):
+        total = 1
+        assert self.fs_map is not None
+        for fs in self.fs_map.get('filesystems'):
+            self.log.info("getting standby_count_wanted")
+            total = max(total, fs.get('mdsmap').get('standby_count_wanted'))
+        return total
+
+    def get_current_standby_count(self):
+        # standbys are not grouped by filesystems in fs_map
+        # available = standby_replay + standby_active
+        assert self.fs_map is not None
+        return len(self.fs_map.get('standbys'))
+
+    def get_current_active_count(self, fs_name):
+        assert self.fs_map is not None
+        for fsys in self.fs_map.get('filesystems'):
+            if fsys.get('mdsmap').get('fs_name') == fs_name:
+                return fsys.get('mdsmap').get('max_mds')
+        return 0
+
+    def get_fs_name(self, index=0):
+        assert self.fs_map is not None
+        fs = self.fs_map.get('filesystems')[index]
+        self.log.info("fs:{}".format(fs))
+        fs_name = fs.get('mdsmap').get('fs_name')
+        self.log.info("fs_name:{}".format(fs_name))
+        return fs_name
+
+    def verify_and_manage_mds_instance(self):
+        standbys_required = self.get_required_standby_count()
+        standbys_current = self.get_current_standby_count()
+        standbys_active = self.get_current_active_count(self.get_fs_name())
+
+        self.log.info("standbys_required:{0}, standbys_current:{1}"
+                      .format(standbys_required, standbys_current))
+        total = standbys_current + standbys_active
+        if standbys_current > standbys_required:
+            # remove one mds at a time and wait for updated fs_map
```

Comment:
Why remove only 1 MDS at a time?

Reply:
Spawning and killing MDS instances is not an instantaneous operation. So, when an MDS instance is spawned, the FSMap is updated per instance and an FSMap update notification is sent to the mgr plugins. So even if the count is bumped up at one go, the mgr plugins get to know the updates one at a time.
eg. if `standby_count_wanted` is bumped up from 1 to 3 and 2 more MDS are requested to be spawned at one go, the mds_autoscaler plugin will get an FSMap update when one instance is spawned and again the calculation to spawn or kill MDS will be done. This time it will fall short by 1 MDS, so again the cephadm will be requested to spawn the missing number of MDS while cephadm is already processing the earlier request.
Now, it may happen that eventually more than the requested MDS might get spawned and the mds_autoscaler plugin will need to decide to kill the extra MDSs.
In this case if we again do a oneshot kill operation, the mds_autoscaler plugin will get FSMap update per MDS instance change and again the the mds_autoscaler plugin will decide to kill extra MDSs in oneshot mode.
To avoid the thrashing, I've decided to bump up and down the MDS instances in single step manner.

Reply:
Because this is driven by fsmap changes, you still have the situation that _any_change in the fsmap (e.g. laggy daemon) will cause this to be rechecked.

If the service spec returned by `update_daemon_count` doesn't change, then I don't think there should be a problem deciding the exact count of daemons desired and then apply that via the orchestrator.

Reply:
@batrick 
yeah, I was worried about other changes to the FSMap as well

okay, this has been fixed now
if `Placement` already has desired count then a `None` is returned from `update_daemon_count()` and the calling function then assumes that there's nothing more that can be done. This works well.